icefoganalytics/travel-authorization
# Travel Desk: Rank Travel Options

Fixes https://github.com/icefoganalytics/travel-authorization/issues/230

Relates to:

- https://github.com/icefoganalytics/travel-authorization/pull/237
- https://github.com/icefoganalytics/travel-authorization/pull/238
- https://github.com/icefoganalytics/travel-authorization/pull/239
- https://github.com/icefoganalytics/travel-authorization/pull/240

# Context

Create the Rank travel desk options screens for Traveler (https://travel-auth-dev.ynet.gov.yk.ca/my-travel-requests/227/request/rank-options).
Pull the grouped segments from travel desk forms across

![Image](https://github.com/user-attachments/assets/b22002ad-43e7-415d-b59e-32a0e06e9f62)

# Implementation

1. Add some design documentation so I could figure out the more of the travel authorization flow.
2. Rebuild the various travel authorization serializers so I can more easily show the correct state dependent screens in the UI.
3. Add some travel agency seeds to make it easier to review.
4. Add a bunch of components around flight option ordering for the traveller.

# Screenshots

Rank Options page
http://localhost:8080/my-travel-requests/5/request/rank-options
![image](https://github.com/user-attachments/assets/54e13f93-9a57-4c35-be65-13c49262d1f4)

Rank Options page with "Does not work" option selected
http://localhost:8080/my-travel-requests/5/request/rank-options
![image](https://github.com/user-attachments/assets/1ec3b26f-afe0-4934-b323-09e0c43a3590)

Awaiting Booking Page
http://localhost:8080/my-travel-requests/4/awaiting-request-booking
![image](https://github.com/user-attachments/assets/71f51b70-ceef-46a2-abc4-6ab681931743)

# Testing Instructions

1. Run the new seeds via
   ```sh
   dev seed run -- --specific 070_travel-agencies-seeds.ts
   ```
2. Run the test suite via `dev test` (or `dev test_api`)
3. Boot the app via `dev up`
4. Log in to the app at http://localhost:8080
5. Click on "My Travel Requests" in the sidebar.
6. Collapse the sidebar via the top left hamburger button.
7. Click on the "New Travel Request" button to create a new travel request.
8. Fill in the "Details: purpose" step, and click "Continue".
9. Fill in the "Details: trip" step, making sure to select "Aircraft" for at least one travel segment, and then click "Continue".
10. Generate estimates in the "Estimate: edit" step, and click "Continue".
11. Review your travel request in the "Details: submit" step.
12. Fill in the "submit to" field with your supervisor's email address, and click "Submit to Supervisor".
13. Review your submitted request in the "Details" review step.
14. Open a private browsing window in your browser and navigate to http://localhost:8080.
15. Log in as the supervisor you submitted the request to.
16. In the sidebar, click on "Manager View".
17. Find your submitted request in the "Pending Approval" card and click on it.
18. Scroll down to the bottom of the page and click "Approve".
19. Agree to the confirmation prompt.
20. Back in your previous browser window, where you are logged in as a regular user click the continue button.
21. Continue pass the estimate review step.
22. On the "Request: traveller details" step, fill in the required fields, and click "Continue".
23. On the "Request: edit" step, check that your flight information looks correct, and click "Submit".
24. Note that you are redirected to the "Request" review step, and that the first half of it is easier to read.
25. Click "Continue" to proceed to the "AWaiting flight options" step.
26. Back in your private browsing window, log in as a travel desk user, click on "Travel Desk" in the sidebar.
27. Find your submitted request in the Travel Desk table and click on the "edit" button on it.
    > Maybe the call to action should go directly to the "manage flight options" page?
28. Scroll down below the Flight Requests table, and click the "Manage Flight Options - Travelport(tm)" button.
29. Paste in the flight info from https://github.com/icefoganalytics/travel-authorization/pull/237 and click "Import Flight Segments".
30. Select all segments, and then click the "group selected" button.
31. In the Add Flight Option dialog, pick the "Leg", and fill in the "Cost" field, and then create the flight option.
32. Repeat the flight option creation process while providing different costs and maybe tinkering with the data a bit.
33. After you create all the flight options you want, scroll down to the bottom of the page, and click the "Return to Travel Request" button.
34. Back in the "Travel Desk Request" page, scroll down to the bottom of the page, and click the "Send to Traveler" button. You will be returned to the "Travel Desk" page.
35. Back in your regular browser window, where you are logged in as a regular user, go back to the "My Travel Requests" page via the breadcrumbs.
36. In the "My Travel" table, find your request, and click the "Rank Travel Desk Options" button.
37. Pick some flight orderings you like, and maybe select "does not work" for one of them.
38. Save each leg ordering as you go, or wait until you submit.
39. After you order all the options, scroll down to the bottom of the page, and click the "Submit Option Rankings" button.
40. You will now see the "Awaiting Booking" page.
41. Go back to the "My Travel Requests" page via the breadcrumbs.
42. Check that you don't have a call to action for your request as there is nothing for you to do yet.
43. Double click on your request in the table, and note that it sends you back to the "Awaiting Booking" page.
